src/analysis/souffle/dataflow_graph.dl

```diff
@@ -73,4 +89,21 @@ path(from, to) :- resolvedEdge(_, from, intermediate), path(intermediate, to).
 isAccessPath(x) :- resolvedEdge(_, x, _).
 isAccessPath(y) :- resolvedEdge(_, _, y).
 
+// The 0th predecessor shall be the predecessor with the smallest src ordinal
+// number.
+orderedPredecessors(ap, for, src, 0) :-
+  resolvedEdge(for, src, ap),
+  ord(src) = min ord(candidate) : { resolvedEdge(for, candidate, ap) }.
+
+// The Nth precessor shall be the precessor with the smallest src ordinal
+// number larger than the N-1th ordinal number.
+orderedPredecessors(ap, for, src, prevOrd + 1) :-
```

Comment:
Will cycles in the graph cause any issue? Do we have test to verify that this is OK?

Reply:
We do not have a test to verify that, I'll add one to make sure.

Reply:
Done.